studies/helpers.py

```diff
@@ -36,3 +41,56 @@ def send_mail(
     )
     email.attach_alternative(html_content, "text/html")
     email.send()
+
+
+class FrameActionDispatcher(object):
+    """Dispatches an action based on the latest frame ID of a given response.
+
+    The idea with this class is that it's easy to extend - you just add a method with the lower-case
+    name matching the frame type you want to act on, and make sure to include the response and frame_data as
+    named arguments, plus optional args/kwargs that can be passed in at call time.
+    """
+
+    def __call__(self, response, *args, **kwargs):
+        """Dispatcher that delegates to inner methods."""
+        if response.sequence:
+            current_frame_id = response.sequence[-1]
+            frame_data = response.exp_data[current_frame_id]
+            frame_type = frame_data.get("frameType", None)
+            if frame_type is None:
+                return
+        else:
+            return
+
+        try:
+            return getattr(self, frame_type.lower())(
+                response, frame_data, *args, **kwargs
+            )
+        except AttributeError as e:
+            logger.warning(f"{e.args} is not registered for frame-specific action.")
```

Comment:
given that the default frame type is 'DEFAULT', want to only warn if it's not registered for frame-specific action and isn't DEFAULT? (Just thinking of readability of these logs.)